### PR TITLE
feat(cip2): Input Selection

### PR DESCRIPTION
**Pull request by mkazlauskas**
# Context

[CIP2](https://cips.cardano.org/cips/cip2/)

# Proposed Solution

Implement Round-Robin Random Improve.

# Important Changes Introduced

- Create new package: `@cardano-sdk/cip2`
- Move BigIntMath from `@cardano-sdk/golden-test-generator` to `@cardano-sdk/core`
- Add some util functions to `cardano-sdk/core`
- Move OgmiosToCardanoWasm to `cardano-sdk/core`

### CODE REVIEW

File: packages/cip2/test/RandomImprove.test.ts [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+//   });
+// };
+
+describe('RandomImprove', () => {
+  let SerializationLib: CardanoSerializationLib;
+  let randomImprove: RandomImprove;
+  let utils: TestUtils;
+  beforeAll(() =>
+    // eslint-disable-next-line promise/always-return
+    loadCardanoSerializationLib().then((lib) => {
```
Comment (rhyslbw) [inline packages/cip2/test/RandomImprove.test.ts:64]: We tend to prefer async/await syntax in most places

Comment (rhyslbw) [inline packages/cip2/test/RandomImprove.test.ts:64]: We tend to prefer async/await syntax in most places

File: packages/cip2/README.md [hunk 2]
```diff
@@ -0,0 +1,10 @@
+# Cardano JS SDK | 2
+
+This package implements concepts from the draft specification being developed in [CIP-0002].
+
+Contains TypeScript types and interfaces, as well as implementation [referenced] from nami-wallet.
+
+TODO: rename coin selection to input selection
+
+[cip-0002]: https://cips.cardano.org/cips/cip2/
+[borrowed]: https://github.com/Berry-Pool/nami-wallet/blob/8d3d2bdfd81b4a16b19df3da978ba578164b663c/src/lib/coinSelection.js
```
Comment (rhyslbw) [inline packages/cip2/README.md:10]:
```suggestion
[referenced]: https://github.com/Berry-Pool/nami-wallet/blob/8d3d2bdfd81b4a16b19df3da978ba578164b663c/src/lib/coinSelection.js
```

File: packages/cip2/package.json [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  "scripts": {
+    "build": "tsc --build ./src",
+    "tsc": "shx echo typescript --noEmit command not implemented yet",
+    "cleanup": "shx rm -rf dist node_modules",
+    "lint": "eslint --ignore-path ../../.eslintignore \"**/*.ts\"",
+    "test": "jest -c ./test/jest.config.js",
+    "test:debug": "DEBUG=true yarn test"
+  },
+  "devDependencies": {
+		"@emurgo/cardano-serialization-lib-nodejs": "^8.0.0",
```
Comment (rhyslbw) [inline packages/cip2/package.json:21]:
```suggestion
    "@emurgo/cardano-serialization-lib-nodejs": "^8.0.0",
```

File: packages/cip2/test/RandomImprove.test.ts [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+import { CardanoSerializationLib } from '@cardano-sdk/cardano-serialization-lib/src/loadCardanoSerializationLib';
+
+const protocolParameters: InputSelectionProtocolParameters = {
+  maxTxSize: 16_384,
+  minFeeCoefficient: 44,
+  minFeeConstant: 155_381,
+  minUtxoValue: 1_000_000
+};
+
+// TODO: Conservation of Outputs and Conservation of UTxO tests are duplicated exactly.
```
Comment (rhyslbw) [inline packages/cip2/test/RandomImprove.test.ts:27]: Yes this is a good refactor

**Review by rhyslbw — CHANGES_REQUESTED** [review]

File: packages/cip2/src/RandomImprove.ts [hunk 5]
```diff
@@ -177,7 +29,8 @@ interface ImproveRange {
 }
 
 /**
- * RandomImprove algorithm implementation for Input Selection
+ * RandomImprove algorithm implementation for Input Selection.
```
Comment (rhyslbw) [inline packages/cip2/src/RandomImprove.ts:33]: :+1:

File: packages/cip2/src/RandomImprove.ts [hunk 6]
```diff
@@ -177,7 +29,8 @@ interface ImproveRange {
...
+ * See more details [here](https://cips.cardano.org/cips/cip2/#randomimprove).
```
Comment (rhyslbw) [inline packages/cip2/src/RandomImprove.ts:34]: :+1:

**Review by rhyslbw — CHANGES_REQUESTED** [review]
Please drop the commits cherry-picked in #87, then rebase on `master`

File: packages/cip2/src/tsconfig.json [hunk 7]
```diff
@@ -0,0 +1,7 @@
+{
+  "extends": "../../../tsconfig.json",
+  "compilerOptions": {
+    "outDir": "../dist/",
+    "rootDir": "."
+  }
```
[the remark this answers is not in the document]
Comment (mkazlauskas) [reply packages/cip2/src/tsconfig.json:6]: Could you please explain the purpose of this? Is it to automatically rebuild `cardano-serialization-lib` package?

[the remark this answers is not in the document]
Comment (rhyslbw) [reply packages/cip2/src/tsconfig.json:6]: Yes it's a [project reference](https://www.typescriptlang.org/docs/handbook/project-references.html)

File: packages/cip2/src/types.ts [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+export interface SelectionResult {
+  selection: {
+    /**
+     * A set of inputs, equivalent to a subset of the initial UTxO set.
+     *
+     * From the point of view of a wallet, this represents the value
+     * that has been selected from the wallet in order to cover the total payment value.
+     */
+    inputs: TransactionUnspentOutput[];
```
Comment (rhyslbw) [inline packages/cip2/src/types.ts:12]:
```suggestion
    utxos: TransactionUnspentOutput[];
```

File: packages/cip2/src/types.ts [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+export type EstimateTxFee = (
+  // Review: should we be using TransactionOutputs type instead?
+  // It's a collection object in CSL.
+  // Same question for InputSelection interface.
+  utxo: TransactionUnspentOutput[],
+  outputs: TransactionOutput[],
+  change: Value[]
+) => bigint;
```
Comment (rhyslbw) [inline packages/cip2/src/types.ts:44]:
```suggestion
) => Promise<number>;
```

Comment (rhyslbw) [reply packages/cip2/src/types.ts:44]: We need to fully construct the transaction in order to estimate the fee, which requires a call to: https://github.com/input-output-hk/cardano-js-sdk/blob/e37fce63b00645897ae1ef6e0220c75ec133d163/packages/wallet/src/KeyManagement/KeyManager.ts#L12-L14

**Review by rhyslbw — CHANGES_REQUESTED** [review]

File: packages/cip2/src/types.ts (hunk 8 above)
Comment (mkazlauskas) [reply packages/cip2/src/types.ts:12]: The [spec](https://cips.cardano.org/cips/cip2/#results) says "A set of **inputs**, equivalent to a subset of the initial UTxO set.". Please confirm this needs to be renamed.

File: packages/cip2/src/types.ts (hunk 9 above)
Comment (mkazlauskas) [reply packages/cip2/src/types.ts:44]: That's unfortunate, it makes the entire `InputSelector.select` async

File: packages/cip2/src/types.ts (hunk 8 above)
Comment (rhyslbw) [reply packages/cip2/src/types.ts:12]: Ok, let's leave it as-is

File: packages/cip2/src/types.ts (hunk 9 above)
Comment (rhyslbw) [reply packages/cip2/src/types.ts:44]: We could also just change the interface, as the in-memory implementation isn't async:

https://github.com/input-output-hk/cardano-js-sdk/blob/master/packages/in-memory-key-manager/src/InMemoryKey.ts#L43-L50

I'll double-check to see what we need to support hardware wallets

Comment (rhyslbw) [reply packages/cip2/src/types.ts:44]: Unfortunately it's an async operation:

https://github.com/cardano-foundation/ledgerjs-hw-app-cardano/blob/47927d42e9b2b0a70463a719cbd745a84cbfce81/src/interactions/signTx.ts#L41-L56

File: packages/cip2/test/util.ts [hunk 10]
```diff
@@ -0,0 +1,209 @@
+// Review: these are utilities used by tests.
```
Comment (rhyslbw) [inline packages/cip2/test/util.ts:1]: Let's leave it here for now, and consider introducing a `util-dev` package

File: packages/cip2/test/util.ts [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+      const amount = output.amount();
+      result = coalesceValueQuantities(result, cslUtils.valueToValueQuantities(amount));
+    }
+    return result;
+  };
+
+  const getTotalInputAmounts = (results: SelectionResult): ValueQuantities =>
+    results.selection.inputs
+      .map((input) => input.output().amount())
+      // eslint-disable-next-line unicorn/no-array-reduce
```
Comment (rhyslbw) [inline packages/cip2/test/util.ts:160]: Also can disable this rule

**Review by rhyslbw — CHANGES_REQUESTED** [review]

File: packages/cip2/test/jest.setup.js [hunk 12]
```diff
@@ -0,0 +1,4 @@
+// TODO: jest environment is not happy with 'lodash-es' exports.
```
[the remark this answers is not in the document]
Comment (mkazlauskas) [reply packages/cip2/test/jest.setup.js:1]: Can this TODO be merged as-is?

[the remark this answers is not in the document]
Comment (rhyslbw) [reply packages/cip2/test/jest.setup.js:1]: Yes

**Comment by mkazlauskas** [comment]
@rhyslbw I think I addressed all the comments, let me know if I missed something. Please take a look at [this](https://github.com/input-output-hk/cardano-js-sdk/pull/84/files#diff-5f90861d94ca998bd5eaf8b0fcc4d8f89446ae8fc72da57e787ec56fe7361292R58) too.

I didn't move the Ogmios<-->CSL value conversion util functions out of cip2 yet, because we need CSL to update lovelace type to bigint first. Left a comment about that.

Also I think we should move OgmiosToCardanoWasm to core package and remove core package dependency from it. Core package has a dependency on CSL package.

**Comment by rhyslbw** [comment]
> Also I think we should move OgmiosToCardanoWasm to core package and remove core package dependency from it. Core package has a dependency on CSL package.

Done in d13513fff3e1ec91cbd4631f348b1bbcf5d0e3a8, which fixes the CI failure

> Please take a look at this too.

I've asked Jonathan for advice. Cardano does have a transaction size limit, so it's safe to assume we need it.

File: packages/core/src/Asset/util.ts [hunk 13]
```diff
@@ -1,33 +1,24 @@
...
+const textDecoder = new TextDecoder();
+const textEncoder = new TextEncoder();
+
+/**
+ * @returns {string} concatenated policy id and asset name
+ */
+export const createAssetId = (scriptHash: ScriptHash, assetName: AssetName): string =>
+  Buffer.from(scriptHash.to_bytes()).toString('hex') + textDecoder.decode(assetName.name());
+
+export const parseAssetId = (id: string, csl: CardanoSerializationLib) => {
```
Comment (rhyslbw) [inline packages/core/src/Asset/util.ts:17]: Let's always us `assetId` to be explicit

File: packages/cip2/src/util.ts [hunk 14]
```diff
@@ -18,61 +18,52 @@ export const transactionOutputsToArray = (outputs: CSL.TransactionOutputs): CSL.
...
+// policy ID and asset name from an asset ID
+export const valueQuantitiesToValue = ({ coins, assets }: ValueQuantities, csl: CardanoSerializationLib): CSL.Value => {
+  const value = csl.Value.new(csl.BigNum.from_str(coins.toString()));
+  if (!assets) {
+    return value;
+  }
+  const assetIds = Object.keys(assets);
+  if (assetIds.length > 0) {
+    const multiasset = csl.MultiAsset.new();
+    for (const id of assetIds) {
```
Comment (rhyslbw) [inline packages/cip2/src/util.ts:59]: `id` -> `assetId`

File: packages/core/src/Asset/util.ts [hunk 15]
```diff
@@ -1,33 +1,24 @@
...
-  const encoder = new TextEncoder();
-  const decoder = new TextDecoder();
+const textDecoder = new TextDecoder();
+const textEncoder = new TextEncoder();
+
+/**
+ * @returns {string} concatenated policy id and asset name
+ */
+export const createAssetId = (scriptHash: ScriptHash, assetName: AssetName): string =>
+  Buffer.from(scriptHash.to_bytes()).toString('hex') + textDecoder.decode(assetName.name());
```
Comment (rhyslbw) [inline packages/core/src/Asset/util.ts:15]: `assetName` shouldn't be decoded. The only time we would decode is for presentation purposes, but there's no enforcement in the ledger so it's not guaranteed to be UTF-8.

**Review by rhyslbw — CHANGES_REQUESTED** [review]

File: packages/cip2/src/util.ts [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+      }
+    }
+  }
+  return result;
+};
+
+// This is equivalent to OgmiosToCardanoWasm.value(), so they should be merged.
+// Note that current implementation of OgmiosToCardanoWasm
+// imports nodejs version of CSL, so it cannot be used in browser.
+// Also note that this function needs to know what bech32 prefix was used in order to recreate
```
Comment (rhyslbw) [inline packages/cip2/src/util.ts:49]: We can hard-code the prefix as `asset` as it's the same across all networks

https://cips.cardano.org/cips/cip5/

File: packages/core/src/Asset/util.ts (hunk 15 above)
Comment (mkazlauskas) [reply packages/core/src/Asset/util.ts:15]: I see. We still need to stringify it. How about `[...assetName].join('')`?

**Review by rhyslbw — APPROVED** [review]
Excellent work @mkazlauskas :tada:

File: packages/core/src/Asset/util.ts (hunk 15 above)
Comment (rhyslbw) [reply packages/core/src/Asset/util.ts]:
```
Buffer.from(asset.name()).toString("hex");
```

**Pull request merged**